File: iosApp/iosApp/Features/AddressCaptureSheet.swift
// AddressCaptureSheet.swift
// Post-authentication address-capture prompt. Surfaces once after the
// user signs in or registers, only when their `delivery_address` is
// blank on the server. Users can save the address or dismiss; either
// path persists a "asked" flag in UserDefaults so we don't pester on
// every cold launch. The Account tab's "Edit profile" entry stays the
// canonical way to update the address later (this sheet just kicks the
// first capture).
//
// Submits via the same `ProfileEditViewModel.save(deliveryAddress:)`
// path as `ProfileEditView`, so the server-side route + KMP plumbing
// for saving is already audited and consistent.

import SwiftUI
import ThapsusShared

struct AddressCaptureSheet: View {
    @Environment(\.dismiss) private var dismiss
    @State private var deliveryAddress: String = ""
    @State private var vm: ProfileEditViewModel? = nil
    @State private var observer: StateFlowObserver<ProfileEditViewModelFormState>? = nil
    @FocusState private var addressFocused: Bool

    /// Called when the user successfully saves or chooses to skip. The
    /// caller is responsible for flipping the persisted "asked" flag so
    /// the sheet isn't re-presented on the next cold launch.
    var onResolved: () -> Void

    var body: some View {
        NavigationStack {
            ZStack {
                AppBackground().ignoresSafeArea()

                ScrollView {
                    VStack(alignment: .leading, spacing: 24) {
                        VStack(alignment: .leading, spacing: 10) {
                            Text("Where should we deliver?")
                                .font(.editorialTitle)
                                .foregroundStyle(Brand.ink)
                                .fixedSize(horizontal: false, vertical: true)
                            Text("Add your Kenya delivery address so our riders know where to drop your parcels. You can change it any time from the Account tab.")
                                .font(.subheadline)
                                .foregroundStyle(Brand.ink.opacity(0.7))
                                .fixedSize(horizontal: false, vertical: true)
                        }

                        VStack(alignment: .leading, spacing: 8) {
                            Text("Kenya delivery address".uppercased())
                                .font(.caption2.weight(.heavy))
                                .tracking(2)
                                .foregroundStyle(Brand.ink.opacity(0.55))
                            TextField(
                                "e.g. Westlands, Nairobi — Apartment 4B",
                                text: $deliveryAddress,
                                axis: .vertical
                            )
                            .lineLimit(3...6)
                            .textFieldStyle(.plain)
                            .focused($addressFocused)
                            .padding(14)
                            .background(
                                RoundedRectangle(cornerRadius: 16, style: .continuous)
                                    .fill(Brand.cream.opacity(0.7))
                            )
                            .overlay(
                                RoundedRectangle(cornerRadius: 16, style: .continuous)
                                    .stroke(Brand.ink.opacity(0.10), lineWidth: 1)
                            )
                            Text("Include landmarks, gate numbers, or anything that helps a rider find you in Nairobi.")
                                .font(.caption2)
                                .foregroundStyle(.secondary)
                        }

                        statusBanner

                        Button(action: save) {
                            Text(isSubmitting ? "Saving…" : "Save address")
                                .font(.headline)
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 4)
                        }
                        .buttonStyle(InkButtonStyle())
                        .disabled(trimmedAddress.isEmpty || isSubmitting)

                        Button(action: skip) {
                            Text("Skip for now")
                                .font(.subheadline.weight(.semibold))
                                .foregroundStyle(Brand.ink.opacity(0.65))
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 12)
                        }
                        .buttonStyle(.plain)
                    }
                    .padding(.horizontal, 22)
                    .padding(.top, 24)
                    .padding(.bottom, 36)
                }
            }
            .scrollDismissesKeyboard(.interactively)
            .navigationTitle("")
            .navigationBarTitleDisplayMode(.inline)
        }
        .presentationDetents([.fraction(0.85), .large])
        .presentationDragIndicator(.visible)
        .interactiveDismissDisabled(false)
        .task { bootstrap() }
        .onChange(of: observer?.value.map { String(describing: type(of: $0)) }) { _, _ in
            if case is ProfileEditViewModelFormStateSaved = observer?.value {
                AppHaptics.fire(.success)
                onResolved()
                dismiss()
            }
        }
        .onAppear { addressFocused = true }
    }

    // MARK: - Helpers

    private var trimmedAddress: String {
        deliveryAddress.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private var isSubmitting: Bool {
        observer?.value is ProfileEditViewModelFormStateSubmitting
    }

    @ViewBuilder
    private var statusBanner: some View {
        if case let err as ProfileEditViewModelFormStateError = observer?.value {
            ErrorBanner(title: "Couldn't save", message: err.message)
        }
    }

    private func save() {
        guard !trimmedAddress.isEmpty else { return }
        vm?.save(
            name: nil,
            phone: nil,
            languagePref: nil,
            deliveryAddress: trimmedAddress
        )
    }

    private func skip() {
        AppHaptics.fire(.tap)
        onResolved()
        dismiss()
    }

    private func bootstrap() {
        guard vm == nil else { return }
        let model = ThapsusSdk.shared.profileEditViewModel()
        vm = model
        observer = StateFlowObserver(initial: model.form.value) {
            model.form
        }
    }
}

File: androidApp/src/main/kotlin/com/thapsus/cargo/android/ui/customer/AddressCaptureSheet.kt
package com.thapsus.cargo.android.ui.customer

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.thapsus.cargo.ThapsusSdk
import com.thapsus.cargo.android.ui.primitives.CalloutBanner
import com.thapsus.cargo.android.ui.primitives.InkButton
import com.thapsus.cargo.android.ui.theme.Brand
import com.thapsus.cargo.presentation.ProfileEditViewModel

// Post-authentication address-capture prompt — Android twin of
// AddressCaptureSheet.swift. ModalBottomSheet expanded to ~85% via the
// default skipPartiallyExpanded behaviour; "Save address" persists via
// the shared ProfileEditViewModel (same path the canonical Edit profile
// flow uses), "Skip for now" closes without writing. Either action
// flips the AddressPromptStore flag so subsequent cold launches don't
// re-show the sheet. The Account tab remains canonical for editing.

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AddressCaptureSheet(
    onResolved: () -> Unit
) {
    val vm = remember { ThapsusSdk.profileEditViewModel() }
    DisposableEffect(vm) { onDispose { vm.clear() } }
    val form by vm.form.collectAsStateWithLifecycle()

    var address by remember { mutableStateOf("") }
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val isSubmitting = form is ProfileEditViewModel.FormState.Submitting

    LaunchedEffect(form) {
        if (form is ProfileEditViewModel.FormState.Saved) {
            onResolved()
        }
    }

    ModalBottomSheet(
        onDismissRequest = { /* dismissal is explicit via Save / Skip — swallow swipe-down */ },
        sheetState = sheetState,
        containerColor = Brand.cream
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 24.dp, vertical = 12.dp)
                .padding(bottom = 24.dp),
            verticalArrangement = Arrangement.spacedBy(18.dp)
        ) {
            Text(
                text = "Where should we deliver?",
                fontSize = 22.sp,
                fontWeight = FontWeight.Bold,
                color = Brand.ink
            )
            Text(
                text = "Add your Kenya delivery address so our riders know where to drop your parcels. You can change it any time from the Account tab.",
                color = Brand.ink.copy(alpha = 0.7f)
            )

            OutlinedTextField(
                value = address,
                onValueChange = { address = it },
                label = { Text("Kenya delivery address") },
                placeholder = { Text("e.g. Westlands, Nairobi — Apartment 4B") },
                minLines = 3,
                maxLines = 6,
                supportingText = {
                    Text("Include landmarks, gate numbers, or anything that helps a rider find you in Nairobi.")
                },
                modifier = Modifier.fillMaxWidth()
            )

            when (val f = form) {
                is ProfileEditViewModel.FormState.Error -> CalloutBanner(
                    title = "Couldn't save",
                    message = f.message
                )
                else -> Unit
            }

            InkButton(
                text = if (isSubmitting) "Saving…" else "Save address",
                enabled = !isSubmitting && address.trim().isNotEmpty(),
                onClick = {
                    vm.save(
                        name = null,
                        phone = null,
                        languagePref = null,
                        deliveryAddress = address.trim()
                    )
                }
            )

            TextButton(
                onClick = onResolved,
                modifier = Modifier.fillMaxWidth(),
                contentPadding = PaddingValues(vertical = 12.dp)
            ) {
                Text(
                    text = "Skip for now",
                    color = Brand.ink.copy(alpha = 0.65f),
                    fontWeight = FontWeight.SemiBold
                )
            }
        }
    }
}
